### ISSUE TITLE
Panic when encoding same batch as parquet across multiple outputs

### ISSUE THREAD

**Issue by nicktelford** [redpanda-data/connect#2885]
If you `fan_out` the same messages to multiple outputs that each try to encode with a `parquet_encode` processor, you get a panic.

```
processor_resources:
  - label: parquet_encoder
    parquet_encode:
      default_compression: zstd
      schema:
        - name: document_id
          type: UTF8

...

output:
  broker:
    outputs:
      - aws_s3:
          batching:
            processors:
              - resource: parquet_encoder
      - aws_s3:
          batching:
            processors:
              - resource: parquet_encoder
```

Panic usually looks like this, but I've seen some other variations:
```
fatal error: concurrent map writes

goroutine 119 [running]:
github.com/redpanda-data/connect/v4/internal/impl/parquet.scrubJSONNumbersObj(...)
        /go/pkg/mod/github.com/redpanda-data/connect/v4@v4.35.3/internal/impl/parquet/util.go:41
github.com/redpanda-data/connect/v4/internal/impl/parquet.scrubJSONNumbers({0x4b92440?, 0xc00232e300})
        /go/pkg/mod/github.com/redpanda-data/connect/v4@v4.35.3/internal/impl/parquet/util.go:30 +0x1ed
github.com/redpanda-data/connect/v4/internal/impl/parquet.(*parquetEncodeProcessor).ProcessBatch(0xc0019f4fa0, {0x3e8?, 0x822a6f?}, {0xc01d13a008, 0x3e8, 0x60?})
        /go/pkg/mod/github.com/redpanda-data/connect/v4@v4.35.3/internal/impl/parquet/processor_encode.go:281 +0x1b8
github.com/redpanda-data/benthos/v4/public/service.(*airGapBatchProcessor).ProcessBatch(0xc001c1cdf0, 0xc018ef0c60, {0xc01afa8008, 0x3e8, 0x3e8})
        /go/pkg/mod/github.com/redpanda-data/benthos/v4@v4.36.0/public/service/processor.go:112 +0xf2
github.com/redpanda-data/benthos/v4/internal/component/processor.(*v2BatchedToV1Processor).ProcessBatch(0xc00319a6c0, {0x61a6f10, 0xc001c08370}, {0xc01afa8008, 0x3e8, 0x3e8})
        /go/pkg/mod/github.com/redpanda-data/benthos/v4@v4.36.0/internal/component/processor/auto_observed.go:223 +0x208
github.com/redpanda-data/benthos/v4/internal/impl/pure.(*resourceProcessor).ProcessBatch.func1({0x6170058?, 0xc00319a6c0?})
        /go/pkg/mod/github.com/redpanda-data/benthos/v4@v4.36.0/internal/impl/pure/processor_resource.go:84 +0x4f
github.com/redpanda-data/benthos/v4/internal/manager.(*Type).AccessProcessor.func1({0x6170058?, 0xc00319a6c0?})
        /go/pkg/mod/github.com/redpanda-data/benthos/v4@v4.36.0/internal/manager/type.go:674 +0x6e
github.com/redpanda-data/benthos/v4/internal/manager.(*liveResource[...]).RAccess(0xc001f792f0?, 0xc001b1fb60?)
        /go/pkg/mod/github.com/redpanda-data/benthos/v4@v4.36.0/internal/manager/live_resources.go:41 +0x98
github.com/redpanda-data/benthos/v4/internal/manager.(*liveResources[...]).RAccess(0x621e7e0, {0xc001b1fb60, 0xf}, 0xc00e019b08)
        /go/pkg/mod/github.com/redpanda-data/benthos/v4@v4.36.0/internal/manager/live_resources.go:155 +0xc6
github.com/redpanda-data/benthos/v4/internal/manager.(*Type).AccessProcessor(0x4eaf200?, {0x458601?, 0x18?}, {0xc001b1fb60?, 0x418798?}, 0x1?)
        /go/pkg/mod/github.com/redpanda-data/benthos/v4@v4.36.0/internal/manager/type.go:669 +0x75
github.com/redpanda-data/benthos/v4/internal/impl/pure.(*resourceProcessor).ProcessBatch(0xc0023a3830, {0x61a6f10, 0xc001c08370}, {0xc01afa8008, 0x3e8, 0x3e8})
        /go/pkg/mod/github.com/redpanda-data/benthos/v4@v4.36.0/internal/impl/pure/processor_resource.go:83 +0x110
github.com/redpanda-data/benthos/v4/internal/component/processor.ExecuteAll({0x61a6f10, 0xc001c08370}, {0xc001e8ba80, 0x2, 0x412245?}, {0xc001f55ca8?, 0x1, 0xc00167b070?})
        /go/pkg/mod/github.com/redpanda-data/benthos/v4@v4.36.0/internal/component/processor/execute.go:19 +0x1c7
github.com/redpanda-data/benthos/v4/internal/batch/policy.(*Batcher).flushAny(0xc000b57110, {0x61a6f10, 0xc001c08370})
        /go/pkg/mod/github.com/redpanda-data/benthos/v4@v4.36.0/internal/batch/policy/policy.go:167 +0x1dd
github.com/redpanda-data/benthos/v4/internal/batch/policy.(*Batcher).Flush(0xc0190e6238?, {0x61a6f10?, 0xc001c08370?})
        /go/pkg/mod/github.com/redpanda-data/benthos/v4@v4.36.0/internal/batch/policy/policy.go:134 +0x30
github.com/redpanda-data/benthos/v4/internal/component/output/batcher.(*Impl).loop(0xc001c46910)
        /go/pkg/mod/github.com/redpanda-data/benthos/v4@v4.36.0/internal/component/output/batcher/batcher.go:128 +0x607
created by github.com/redpanda-data/benthos/v4/internal/component/output/batcher.(*Impl).Consume in goroutine 1
        /go/pkg/mod/github.com/redpanda-data/benthos/v4@v4.36.0/internal/component/output/batcher/batcher.go:178 +0xb6
```

My guess is that the processor isn't thread-safe and is being shared when used as a resource. I assumed the resource would construct an independent, duplicate, processor for each output.

**Comment by nicktelford** [comment]
Actually I may be wrong about the root-cause here, because it still occurs even when duplicating the `parquet_encode` processor explicitly.

**Comment by nicktelford** [comment]
OK, so it looks like the _real_ cause here is that `parquet_encode` mutates the documents in the output batch, and I think that batch is shared among the output processors of a `broker`? So when you try to `fan_out` to multiple outputs, they each try to mutate the same documents concurrently, causing a panic.

That's my understanding from: https://github.com/redpanda-data/connect/blob/main/internal/impl/parquet/processor_encode.go#L281

**Comment by mihaitodor** [comment]
Hey @nicktelford, thank you for reporting this bug! Would you be able to try out the proposed fix from #2888? I wasn't able to reproduce the panic you're seeing in my simple unit test, but it does trigger an error without that fix.
